xds/internal/balancer/cdsbalancer/cdsbalancer_test.go

```diff
@@ -315,507 +322,1272 @@ func setupWithWatch(t *testing.T) (*fakeclient.Client, *cdsBalancer, *testEDSBal
-	defer func() {
-		cancel()
-		cdsB.Close()
-	}()
-	// Here we invoke the watch callback registered on the fake xdsClient. This
-	// will trigger the watch handler on the CDS balancer, which will update
-	// the service's counter with the new max requests.
-	var maxRequests uint32 = 1
-	cdsUpdate := xdsresource.ClusterUpdate{
-		ClusterName: clusterName,
-		MaxRequests: &maxRequests,
-		LBPolicy:    wrrLocalityLBConfigJSON,
+	// Verify that the resolver error is pushed to the child policy.
+	select {
+	case err := <-resolverErrCh:
+		if err != resolverErr {
+			t.Fatalf("Error pushed to child policy is %v, want %v", err, resolverErr)
+		}
+	case <-ctx.Done():
+		t.Fatal("Timeout when waiting for resolver error to be pushed to the child policy")
 	}
-	wantCCS := edsCCS(clusterName, &maxRequests, false, wrrLocalityLBConfigJSON, noopODLBCfgJSON)
-	ctx, ctxCancel := context.WithTimeout(context.Background(), defaultTestTimeout)
-	defer ctxCancel()
-	if err := invokeWatchCbAndWait(ctx, xdsC, cdsWatchInfo{cdsUpdate, nil}, wantCCS, edsB); err != nil {
-		t.Fatal(err)
+
+	// Push a resource-not-found-error this time around.
+	resolverErr = xdsresource.NewErrorf(xdsresource.ErrorTypeResourceNotFound, "xds resource not found error")
+	r.ReportError(resolverErr)
+
+	// Wait for the CDS resource to be not requested anymore.
+	select {
+	case <-ctx.Done():
+		t.Fatal("Timeout when waiting for CDS resource to not requested")
+	case <-cdsResourceCanceledCh:
 	}
 
-	// Since the counter's max requests was set to 1, the first request should
-	// succeed and the second should fail.
-	counter := xdsclient.GetClusterRequestsCounter(clusterName, "")
-	if err := counter.StartRequest(maxRequests); err != nil {
-		t.Fatal(err)
+	// Verify that the resolver error is pushed to the child policy.
+	select {
+	case err := <-resolverErrCh:
+		if err != resolverErr {
+			t.Fatalf("Error pushed to child policy is %v, want %v", err, resolverErr)
+		}
+	case <-ctx.Done():
+		t.Fatal("Timeout when waiting for resolver error to be pushed to the child policy")
 	}
-	if err := counter.StartRequest(maxRequests); err == nil {
-		t.Fatal("unexpected success on start request over max")
+
+	// Ensure that the ClientConn moves to TransientFailure.
+	for state := cc.GetState(); state != connectivity.TransientFailure; state = cc.GetState() {
+		if !cc.WaitForStateChange(ctx, state) {
+			t.Fatalf("Timed out waiting for state change. got %v; want %v", state, connectivity.TransientFailure)
+		}
 	}
-	counter.EndRequest()
 }
 
-// TestClose verifies the Close() method in the CDS balancer.
+// Tests that closing the cds LB policy results in the cluster resource watch
+// being cancelled and the child policy being closed.
 func (s) TestClose(t *testing.T) {
-	grpctest.TLogger.ExpectError("cds-lb.*Received balancer config after close")
+	// Register a wrapped cds LB policy for the duration of this test that
+	// allows explicitly closing the LB policy.
```

Comment:
Yeah, the only way to close it would be to close the channel. But, now I have moved all the wrapped registrations to helpers. So, shouldn't be an issue I would assume.

Reply:
Yeah, forgot this was an e2e test :).